### PR TITLE
Add an option for creating Machines using InstanceTemplate.

### PR DESCRIPTION
**Pull request by oxddr**
**What this PR does / why we need it**:

Add an option to create Machine objects using given InstanceTemplate. 

**Special notes for your reviewer**:

This is my first bigger change to Kubernetes-related project, so please be extra careful - I might have implement things completely wrong. Also feel free to be extra picky - so I can learn even more from a review. 

I am not sure what tests to add, so please point it to me. It was hard for me to induct this from current code.

### CODE REVIEW

File: pkg/apis/gceproviderconfig/v1alpha1/gcemachineproviderconfig_types.go [hunk 1]
```diff
@@ -32,11 +32,13 @@ type GCEMachineProviderConfig struct {
 	Roles []MachineRole `json:"roles,omitempty"`
 
 	Zone        string `json:"zone"`
-	MachineType string `json:"machineType"`
+	MachineType string `json:"machineType,omitempty"`
```
Comment (maisem) [inline pkg/apis/gceproviderconfig/v1alpha1/gcemachineproviderconfig_types.go]: Can we also add project here?

File: pkg/cloud/google/machineactuator.go [hunk 2]
```diff
@@ -255,93 +255,109 @@ func (gce *GCEClient) Create(cluster *clusterv1.Cluster, machine *clusterv1.Mach
...
+	if gce.client != nil {
+		return gce.updateAnnotations(cluster, machine)
+	}
+	return nil
+}
+
+func (gce *GCEClient) createFromInstanceTemplate(cluster *clusterv1.Cluster, clusterConfig *gceconfigv1.GCEClusterProviderConfig, machine *clusterv1.Machine, machineConfig *gceconfigv1.GCEMachineProviderConfig) (*compute.Operation, error) {
+	name := machine.ObjectMeta.Name
+	project := clusterConfig.Project
+	zone := machineConfig.Zone
```
Comment (maisem) [inline pkg/cloud/google/machineactuator.go]: inline?

File: pkg/cloud/google/machineactuator.go [hunk 3]
```diff
@@ -255,93 +255,109 @@ func (gce *GCEClient) Create(cluster *clusterv1.Cluster, machine *clusterv1.Mach
 		return gce.handleMachineError(machine, verr, createEventAction)
 	}
```
Comment (maisem) [inline pkg/cloud/google/machineactuator.go]: Can you also move the machineSetupConfigGetter checks to create? If an instanceTemplate is provided we don't need to rely on machineSetupConfigs.

File: pkg/cloud/google/machineactuator.go [hunk 4]
```diff
@@ -255,93 +255,109 @@ func (gce *GCEClient) Create(cluster *clusterv1.Cluster, machine *clusterv1.Mach
...
```
Comment (maisem) [inline pkg/cloud/google/machineactuator.go]: If we add a project field to machine provider spec we can remove the reliance on the cluster object as well if the machine was created via an instanceTemplate.

File: pkg/apis/gceproviderconfig/v1alpha1/gcemachineproviderconfig_types.go (hunk 1 above)
Comment (oxddr) [reply pkg/apis/gceproviderconfig/v1alpha1/gcemachineproviderconfig_types.go]: Project is part of cluster config. I don't think we need to duplicate this here. Or do you have another use case in mind?

File: pkg/cloud/google/machineactuator.go (hunk 4 above)
Comment (oxddr) [reply pkg/cloud/google/machineactuator.go]: I recall `project` was part of machine config once, but it was removed for some reasons (I don't know them, code/config duplication maybe?). In all other places (e.g. in machine actuator), reconciliation of Cluster API relies on `Cluster` anyway.

**Comment by oxddr** [comment]
Friendly ping :)

**Comment by oxddr** [comment]
/assign @justinsb  

Could you please review this PR?

**Comment by detiber** [comment]
@oxddr are you still interested in this change?

**Comment by oxddr** [comment]
Not personally, this could be useful change though. I fine with both closing it or having rebased and possible merged.

**Comment by detiber** [comment]
@oxddr ok, I'm going to close and we can potentially revisit after the v1alpha2 changes land.
/close

**Pull request closed without merging**